src/gluonnlp/utils/parameter.py

```diff
@@ -172,14 +224,20 @@ def save_states(trainer, fname):
     `optimizer.param_dict`, which contains Parameter information (such as
     `lr_mult` and `wd_mult`) will not be saved.
     """
-    if C.S3_PREFIX in fname:
-        # create temp dir
-        temp_dir = os.path.join(tempfile.gettempdir(), str(hash(os.times())))
-        if not os.path.exists(temp_dir):
-            os.makedirs(temp_dir)
-        temp_path = os.path.join(temp_dir, str(hash(os.times())))
-        trainer.save_states(temp_path)
-        _upload_file(temp_path, fname)
-        os.remove(temp_path)
-    else:
-        trainer.save_states(fname)
+    _s3_compatible_save_load(True, trainer.save_states, fname)
+
+class TempFilePath(object):
```

Comment:
it seems weird to have temp file path in parameter.py. Maybe either create a new module file in utils or make this class internal with the `_` prefix